compiler/test/suites/modules.re

```diff
@@ -0,0 +1,116 @@
+open Grain_tests.TestFramework;
+open Grain_tests.Runner;
+
+describe("modules", ({test, testSkip}) => {
+  let test_or_skip =
+    Sys.backend_type == Other("js_of_ocaml") ? testSkip : test;
+
+  let assertSnapshot = makeSnapshotRunner(test);
+  let assertCompileError = makeCompileErrorRunner(test);
+  let assertRun = makeRunner(test_or_skip);
+  let assertFileRun = makeFileRunner(test_or_skip);
+
+  assertSnapshot(
+    "smallest_module",
+    {|
+      module Foo {
+        void
+      }
+    |},
+  );
+  assertRun(
+    "module_side_effect",
+    {|
+      module Foo {
+        print("foo")
+      }
+    |},
+    "foo\n",
+  );
+  assertRun(
+    "module_provide_value",
+    {|
+      module Foo {
+        provide let foo = "foo"
+      }
+      print(Foo.foo)
+    |},
+    "foo\n",
+  );
+  assertRun(
+    "module_provide_function",
+    {|
+      module Foo {
+        provide let foo = () => "foo"
+      }
+      print(Foo.foo())
+    |},
+    "foo\n",
+  );
+  assertRun(
+    "module_provide_type",
+    {|
+      module Foo {
+        provide enum Foo { Foo }
+      }
+      print(Foo.Foo)
+    |},
+    "Foo\n",
+  );
+  assertRun(
+    "module_provide_module",
+    {|
+      module Foo {
+        provide module Foo {
+          provide let foo = () => "foo"
+        }
+      }
+      print(Foo.Foo.foo())
+    |},
+    "foo\n",
+  );
+  assertRun(
+    "module_shadow",
+    {|
+      module Foo {
+        provide let foo = "foo"
+      }
+      module Foo {
+        provide let foo = "bar"
+      }
+      print(Foo.foo)
+    |},
+    "bar\n",
+  );
+  assertRun(
+    "local_module_use",
+    {|
+      module Foo {
```

Comment:
How about a test like:
```
let foo = "foo"

module Foo {
  provide let foo = foo
}

print(Foo.foo)
```
to show that we have everything defined above the module within the scope.

Reply:
Added in 0eacd824